### ISSUE TITLE
member_id file is empty, node won't join cluster

### ISSUE THREAD

**Issue by justinrush** [bitnami/bitnami-docker-etcd#39]
### Name and Version

bitnamic/etcd 3.5.1-debian-10-r88

### What steps will reproduce the bug?

I'm able to reproduce this every ~5ish times using a minimal config with replicas = 5 and using the helm chart. I suspect replicas at 3 will encounter the same problem.

```sh
cat << EOF >values.yaml
auth:
  rbac:
    enabled: false
    create: false
    allowNoneAuthentication: true
replicaCount: 5
EOF

kubectl create ns test
helm install test-etcd bitnami/etcd -n test -f values.yaml
```

Given what code is being executed, the problem appears to be in the `libetcd.sh` code in this repository rather than in the helm chart.

### What is the expected behavior?

`/bitnami/etcd/data/member_id` should always be populated with a hexadecimal string. 

### What do you see instead?

Sometimes this file is written empty.

If you wait for the pods to all be up - not necessarily ready, but up, you can run this to get the contents of member_id:
`for i in `seq 0 4`; do echo $i; kubectl -n test exec -ti test-etcd-${i} cat /bitnami/etcd/data/member_id; echo ""; done`

The pods take a few seconds to put data in there. If the pod isn't there yet, the error will say the file doesn't exist. However, if you get nothing back, then that means you have encountered this error situation.

For example:
```
0
560b99a165a52b82

1
c80222ff4a10c54a

2
14370eaa996d304b

3

4
150ba3cc7eb96aaa

```

The pod in question will start spewing errors and eventually crash due to the readiness probe failing. Errors seem to be in two buckets:

```
{"level":"warn","ts":"2022-03-15T16:05:34.148Z","caller":"rafthttp/http.go:500","msg":"request cluster ID mismatch","local-member-id":"86d0b5c627b713cd","local-member-cluster-id":"f362f0133812732b","local-member-server-version":"3.5.1","local-member-server-minimum-cluster-version":"3.0.0","remote-peer-server-name":"560b99a165a52b82","remote-peer-server-version":"3.5.1","remote-peer-server-minimum-cluster-version":"3.0.0","remote-peer-cluster-id":"c6bc1e028247e245"}
```

and

```
{"level":"warn","ts":"2022-03-15T16:05:34.222Z","caller":"rafthttp/stream.go:653","msg":"request sent was ignored by remote peer due to cluster ID mismatch","remote-peer-id":"150ba3cc7eb96aaa","remote-peer-cluster-id":"c6bc1e028247e245","local-member-id":"86d0b5c627b713cd","local-member-cluster-id":"f362f0133812732b","error":"cluster ID mismatch"}
```

When the pod starts back up, it will detect that [the member_id file exists](https://github.com/bitnami/bitnami-docker-etcd/blob/master/3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh#L664) and attempt to use its the contents of this empty file as an integer and fail with an [error](https://github.com/etcd-io/etcd/blob/f2bc5eee916de63ac525ce5a843e849f73921415/etcdctl/ctlv3/command/member_command.go#L207):
```
etcd 14:52:32.07 INFO  ==> Detected data from previous deployments
etcd 14:52:32.24 INFO  ==> Updating member in existing cluster
Error: bad member ID arg (strconv.ParseUint: parsing "": invalid syntax), expecting ID in Hex
```

I haven't fully parsed through what the `libetcd.sh` file does to confirm, but I highly suspect there are dragons [here](https://github.com/bitnami/bitnami-docker-etcd/blob/master/3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh#L315) - either related to the flushing to disk or something else preventing the data from being written to the `member_id` file.

### Additional information

We are running on AKS if that matters. This seems related to #29 as well as [this issue](https://github.com/bitnami/charts/issues/5771) on the helm chart repository.

I'm hoping someone with knowledge of what this large shell script does and why it does it will know where to add more checks to make sure this doesn't happen.

**Comment by justinrush** [comment]
That explains why it seems to happen more frequently with 5 node clusters vs 3. Is there a deterministic approach rather than a "sleep" approach that is valid?

Why timeout at all? why not let it continue forever and rely on the orchestrator to handle the not-ready condition (be it docker compose or kubernetes)

Happy to do a PR

**Comment by marcosbc** [comment]
Hi @justinrush, the intention of the timeout is to end up in an error state and not leave a container stuck indefinitely at that point. While in Kubernetes the pod would eventually die due to health probes, it is not the case in other environments (like docker-compose or clustered VMs).

If you have any ideas to do it in a deterministic way, feel free to share that idea and/or contribute it. However the simplest way here is to apply a timeout and is why we followed that approach, as the other approach was considered to be too expensive in terms of code complexity and extra config required. Maybe increasing it should be enough.

That said, it could be worth to improve the logic to indicate somehow that the persisted installation had not finished to initialize, and causing an error before it is started up, so it is clear why the node bootstrap actually failed. At that point it could be worth adding an extra flag/envvar to force a re-initialization of all failed environments (i.e. by removing all persisted files for that specific pod).

**Comment by justinrush** [comment]
I'm working on fixing this, just got back from holiday. Keeping it open.

**Comment by cunningr** [comment]
I think we are also hitting this when running `kubectl delete pod/etcd-testing-2` to test recovery.

```
etcd 13:19:35.97
etcd 13:19:35.97 Welcome to the Bitnami etcd container
etcd 13:19:35.97 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-etcd
etcd 13:19:35.97 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-etcd/issues
etcd 13:19:35.97
etcd 13:19:35.97 INFO  ==> ** Starting etcd setup **
etcd 13:19:35.99 INFO  ==> Validating settings in ETCD_* env vars..
etcd 13:19:35.99 INFO  ==> Initializing etcd
etcd 13:19:35.99 INFO  ==> Generating etcd config file using env variables
etcd 13:19:36.01 INFO  ==> Detected data from previous deployments
etcd 13:19:36.49 INFO  ==> Updating member in existing cluster
Error: bad member ID arg (strconv.ParseUint: parsing "": invalid syntax), expecting ID in Hex
```

**Comment by marcosbc** [comment]
@ingvar-goryainov Have you taken a look at the [previous comment with a possible solution](https://github.com/bitnami/bitnami-docker-etcd/issues/39#issuecomment-1069322973)? You may be interested in applying that as a temporary workaround until this is solved. It seems like @justinrush is working on the fix within a PR, but it is taking a bit longer than expected.

**Comment by justinrush** [comment]
I've got a change in and am testing it today

**Comment by justinrush** [comment]
@cunningr - was the pod you deleted in your test actually ready? It looks like it never wrote  the member_id file. I would have expected a delete of the pod to start back up and read the existing member_id file.

I was trying to fix this by testing for member_id being empty and deleting it on start up if thats the case. I don't really want to increase the sleep command because its not really solving the problem, its just working around it for some instances.

The problem is that when it fails to write `member_id`, it may have still written some data. In the case where it never gets a member id, the data is useless. The way I've worked around this in my testing is to add this before the large if block in `etcd_initialize` in `3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh`

```sh
    # initial population of the member_id file can result in an empty file
    # details: https://github.com/bitnami/bitnami-docker-etcd/issues/39
    # if we have an empty member_id file, thats invalid and we should delete it
    if [ -s ${ETCD_DATA_DIR}/member_id ]; then
        info "existing member_id: $(cat ${ETCD_DATA_DIR}/member_id)"
    else
        info "Detected empty member_id, deleting all data"
        rm -rf ${ETCD_DATA_DIR}/*
    fi
```

I've verified this solves the problem I'm trying to solve, but I don't know if this is an acceptable solution for all use cases... `rm -rf`'ing a database's data dir isn't the most graceful thing in the world.  That said, if there is no valid member_id present, the pod will never join the cluster, so it doesn't really matter anyway - to get it to rejoin you will need to restore from backup. Feels like the worst case is that the `member_id` file somehow gets corrupted and there is good data on that pod. The start up script will end up rm -rf'ing the data, which will be re-pulled form other members.

If this seems like an OK approach, I can submit the PR tomorrow

**Comment by cunningr** [comment]
@justinrush - sorry for the delay.  I think my issue was actually self inflicted.  I found that I had the `initialClusterState: existing` explicitly set when deploying a new cluster.

**Comment by HelloKevinTian** [comment]
**Issue**
I have the same issue. I met this when I install apisix followed by [APISIX-INSTALL-README](https://github.com/apache/apisix-ingress-controller/blob/master/docs/en/latest/deployments/kubesphere.md).

Waiting for the PR.

Well, Is there a temporary solution for the issue?

**Version Info**
etcd:3.4.18-debian-10-r14
kubernetes v1.19.12
kubesphere v3.2.1

**Error Info In Pod apisix-etcd-0**
` etcd 10:07:36.04 
 etcd 10:07:36.04 Welcome to the Bitnami etcd container
 etcd 10:07:36.04 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-etcd
 etcd 10:07:36.04 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-etcd/issues
 etcd 10:07:36.04 
 etcd 10:07:36.04 INFO  ==> ** Starting etcd setup **
 etcd 10:07:36.06 INFO  ==> Validating settings in ETCD_* env vars..
 etcd 10:07:36.06 WARN  ==> You set the environment variable ALLOW_NONE_AUTHENTICATION=yes. For safety reasons, do not use this flag in a production environment.
 etcd 10:07:36.07 INFO  ==> Initializing etcd
 etcd 10:07:36.07 INFO  ==> Generating etcd config file using env variables
 etcd 10:07:36.09 INFO  ==> Detected data from previous deployments
 etcd 10:07:36.23 INFO  ==> Updating member in existing cluster
 Error: bad member ID arg (strconv.ParseUint: parsing "": invalid syntax), expecting ID in Hex`

**Extra Info**
In addition, Pod apisix-etcd-1  and apisix-etcd-2 id ok.

**Comment by justinrush** [comment]
You can increase the timeout in the script and re-build the container image and use that image and its likely that will fix your problem. If the approach I mention above is acceptable I can submit a PR to do it.

**Comment by haorenfsa** [comment]
Guys, I really doubt the necessarity of the 2 things the scripts are trying to do:
- maintaining a `member_id`
- the self removal from cluster when a pod shutdown

First you can see from this issue & the all the links to it, that there's a quite big chance that the files keeping track of these two things failed to sync to disk.

The member_id is persisted inside etcd's storage. So there's no reason to track it externally.

About the self removal, it only necessary when you want to shrink the size of cluster. While even in the shrinking situation, it's still not a good idea to do this inside an etcd pod. The prestop hook is not promised. Leave it to an external deployment in the helm chart may be a better idea.

So let's get rid of the `member_id` & `self removal` mechanism.

**Comment by cunningr** [comment]
Will this still cover use cases without persistent storage? We are running
a cluster with 5 nodes and no persistent storage so whenever a pod
goes/starts, it needs to join as a new node. I believe for this case, we
need to have removed the old node first.

**Comment by haorenfsa** [comment]
> Will this still cover use cases without persistent storage? We are running a cluster with 5 nodes and no persistent storage so whenever a pod goes/starts, it needs to join as a new node. I believe for this case, we need to have removed the old node first.

I just tested this case, it can still join by itself. The endpoint of the pod is not changed, so it can still be recognized by the rest nodes as the old one, as long as you have more than a half nodes alive.

**Comment by haorenfsa** [comment]
While the above change is quite big. Let's try a simpler approach first:

We should using `sync -d $filename` rather than the `stdbuf -oL` that we're using now. 

The latter only guarantees the data in stdout buffer is passed to next process in pipeline line by line, while the former ensures the file data is synced to disk (which is the root cause of this case).

I strongly suggest calling sync after every file I/O, as long as the file is going to matter in the restart procedure.

REF: 
[sync command](ref: https://man7.org/linux/man-pages/man1/sync.1.html)
[stdbuf command](https://man7.org/linux/man-pages/man1/stdbuf.1.html)

**Comment by haorenfsa** [comment]
Time to end the disk syncing issue 💪

**Comment by croomes** [comment]
For anyone coming across this issue and just wanting to recover without building new images:
1. Find the `member_id` of the Pod by looking in the logs of an active member Pod, e.g.:
    ````
    2022-05-27 11:45:13.552206 W | rafthttp: health check for peer a5dbbc42933201b6 could not connect: dial tcp 10.244.1.71:2380: connect: connection refused
    ```
2. Find the uid (e.g. `35a21967-d62f-413c-a909-0e5cc14a880a`) and node of the failing pod.
3. Find the PV name of the data volume attached (e.g. `pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b`)
4. Get a shell on the node (https://raw.githubusercontent.com/andyzhangx/demo/master/dev/kubectl-enter)
5. Verify that the `member_id` file is empty:
   ```
    root@aks-agentpool-24495340-vmss000001:/# ls -ltr /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id 
    -rwx------ 1 1001 1001  0 May 27 11:47 /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id
    ```
6. Add the member_id to the file:
    ```
    echo "a5dbbc42933201b6" > /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id
    ```
The Pod should now start correctly.

**Comment by justinrush** [comment]
still an issue

**Comment by haorenfsa** [comment]
> still an issue

Try out our image built with the patch #42 which solves it totally if you're interested `milvusdb/etcd:3.5.4-r1`

**Comment by haorenfsa** [comment]
> For anyone coming across this issue and just wanting to recover without building new images:
> 
> 1. Find the `member_id` of the Pod by looking in the logs of an active member Pod, e.g.:
>    ```
>    2022-05-27 11:45:13.552206 W | rafthttp: health check for peer a5dbbc42933201b6 could not connect: dial tcp 10.244.1.71:2380: connect: connection refused
>    ```
>    ```
> 2. Find the uid (e.g. `35a21967-d62f-413c-a909-0e5cc14a880a`) and node of the failing pod.
> 3. Find the PV name of the data volume attached (e.g. `pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b`)
> 4. Get a shell on the node (https://raw.githubusercontent.com/andyzhangx/demo/master/dev/kubectl-enter)
> 5. Verify that the `member_id` file is empty:
>    ```
>     root@aks-agentpool-24495340-vmss000001:/# ls -ltr /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id 
>     -rwx------ 1 1001 1001  0 May 27 11:47 /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id
>    ```
> 6. Add the member_id to the file:
>    ```
>    echo "a5dbbc42933201b6" > /var/lib/kubelet/pods/35a21967-d62f-413c-a909-0e5cc14a880a/volumes/kubernetes.io~csi/pvc-f3d4ee27-71f5-4a4a-b9c5-2a6df9bb635b/mount/data/member_id
>    ```
> 
> The Pod should now start correctly.

If there's only one pod down in your cluster of 3 or more nodes. An easier way is just delete the pvc, and delete the pod to release the pvc and relaunch the new pvc & pod

Only need 2 lines commands:

```
kubectl -n $namepace delete pvc data-${pod-name} --wait=false
kubectl -n $namepace delete pod ${pod-name}
```

**Comment by ondrej-smola** [comment]
should be fixed by #42
